src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProvider.java

```diff
@@ -0,0 +1,214 @@
+     * the provided reference element. For example, if both are elements are in the same file, there's no reason to
+     * include the target element's file name in the description, but if they're in different files, it's useful to
+     * know where the target element resides.
+     *
+     * @param element          the element for which a description should be returned
+     * @param referenceElement the reference element from which the description is being requested
+     * @return the element description
+     */
+    @Nullable
+    @ApiStatus.Internal
+    public String getElementDescription(@NotNull PsiElement element, @Nullable PsiElement referenceElement) {
+        if (findElementAt(element.getTextOffset()) instanceof LSPSemanticTokenPsiElement semanticTokenElement) {
+            return semanticTokenElement.getDescription(referenceElement);
+        }
+        return null;
+    }
+
+    // Semantic tokens management
+
+    // Store the file's semantic tokens so that we have constant-time lookup of an element for a given offset
+    @Nullable
+    private Map<Integer, LSPSemanticToken> getSemanticTokensByOffset() {
+        PsiFile file = getFile();
+        if (file == null) return null;
+
+        // By caching the storage on the file this way, it's automatically evicted when the file changes
+        return CachedValuesManager.getCachedValue(file, new CachedValueProvider<>() {
+            @Override
+            @NotNull
+            public Result<Map<Integer, LSPSemanticToken>> compute() {
+                Map<Integer, LSPSemanticToken> semanticTokensByOffset = Collections.synchronizedMap(new HashMap<>());
+                return Result.create(semanticTokensByOffset, file);
+            }
+        });
+    }
+
+    /**
+     * Adds a semantic token for the file.
+     *
+     * @param textRange      the token's text range in the file
+     * @param tokenType      the token type
+     * @param tokenModifiers the token modifiers
+     */
+    @ApiStatus.Internal
+    public void addSemanticToken(@NotNull TextRange textRange,
+                                 @Nullable String tokenType,
+                                 @Nullable List<String> tokenModifiers) {
+        PsiFile file = getFile();
+        if (file == null) return;
+
+        Map<Integer, LSPSemanticToken> semanticTokensByOffset = getSemanticTokensByOffset();
+        if (semanticTokensByOffset != null) {
+            LSPSemanticToken semanticToken = new LSPSemanticToken(file, textRange, tokenType, tokenModifiers);
+
+            // Index the token for its text range
+            for (int offset = textRange.getStartOffset(); offset <= textRange.getEndOffset(); offset++) {
+                semanticTokensByOffset.put(offset, semanticToken);
+            }
+        }
+    }
+
+    @Nullable
+    private LSPSemanticToken getSemanticToken(int offset) {
+        PsiFile file = getFile();
```

Comment:
I have not tested your PR, but I wonder if this method is called a lot. My fear is to create a lot of fake PsiElement as soon as we type a character in the editor.